x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/MapReduceAggregator.java

```diff
@@ -0,0 +1,154 @@
+import org.elasticsearch.search.aggregations.AggregatorBase;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.CardinalityUpperBound;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.xcontent.ToXContent;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.ml.aggs.mapreduce.MapReduceValueSourceRegistry.REGISTRY_KEY;
+
+public abstract class MapReduceAggregator<MapReduceContext extends Writeable & Closeable, MapReduceResult extends ToXContent & Writeable>
+    extends AggregatorBase {
+
+    private final List<ValuesExtractor> extractors;
+    private final AbstractMapReducer<MapReduceContext, MapReduceResult> mapReducer;
+    private final BigArrays bigArraysForMapReduce;
+    private final LongObjectPagedHashMap<MapReduceContext> mapReduceContextByBucketOrdinal;
+    private final boolean profiling;
+    private final DelegatingCircuitBreakerService breakerService;
+
+    protected MapReduceAggregator(
+        String name,
+        AggregationContext context,
+        Aggregator parent,
+        Map<String, Object> metadata,
+        AbstractMapReducer<MapReduceContext, MapReduceResult> mapReducer,
+        List<ValuesSourceConfig> configs
+    ) throws IOException {
+        super(name, AggregatorFactories.EMPTY, context, parent, CardinalityUpperBound.NONE, metadata);
+
+        this.extractors = configs.stream()
+            .map(c -> context.getValuesSourceRegistry().getAggregator(REGISTRY_KEY, c).build(c))
+            .collect(Collectors.toList());
+
+        this.mapReducer = mapReducer;
+        this.profiling = context.profiling();
+        this.mapReduceContextByBucketOrdinal = new LongObjectPagedHashMap<>(1, context.bigArrays());
+
+        // big arrays used for the map reduce context have a lifespan beyond this aggregator, so they can't use the bigarray from the
+        // context. But the non-recycling big array instance does not trip the circuit breaker, so we do a trick here:
+        // We use the non-recycling big array but with the circuit breaker of the context and the counter of the aggregation.
+        // This ensures that we 1st of all have a circuit breaker _and_ that all bytes that are added to the circuit breaker get deducted
+        // later. This is important as otherwise the circuit breaker continues with a wrong count and eventually trips.
+        // (if we would use the big array breaker we would not deduct the counter)
```

Comment:
I am not understanding this fully and what I am understanding seems very dangerous.

The `buildAggregations` may or may not serialize to another node (and thus serialize to another circuit breaker)

Once `buildAggregaions` is called (the end goal of the aggregator), this aggregator is closed. The fact that things may "live on" after closing seems really dangerous. 

Where do we call `.close()` on `TransactionStore` as that is the time when the circuit breaker will have the bytes returned to it? From what I can tell that only happens on reduce? If that is the case, we will leak bytes in teh circuit breaker as the bytes reserved for the `TransactionStore` on the data node are never returned.

Reply:
it uses the _non-leaking_  `NON_RECYCLING_INSTANCE`, that means _no_ page cache recycler. This shouldn't/can't leak memory. I attach the circuit breaker of the agg context, to circuit break and _not_ produce an OOM.

The dangerous part: I absolutely must end up at `0` at the end. Everything I add to the cb, I must substract. The CB is global, so if I e.g. forget to substract `X` bytes from the cb, the cb will at some point _think_ that it runs out of memory. Somewhat a "virtual memory leak".

To work around I am using the agg context `addRequestCircuitBreakerBytes`, because if you drill down into it, you will see that it remembers everything it added and will send `-requestBytesUsed` at the end of its lifetime and therefore bring balance.

Reply:
for a short moment however the memory got substracted from the cb, but memory isn't really freed, because the context(transactionstore) is still there. Same for the HLL++ in cardinality. That is my current understanding. The deep-copy solution suffers the same problem, so switching to that won't help.

(I will of course run many tests to find out that this works properly)

Reply:
I see how the circuit breaker works now. We effectively add to it while in use, `doClose` is called, and we disconnect it (effectively making it a no-op for the `NON_RECYCLING_INSTANCE`) while the `NON_RECYCLING_INSTANCE` objects are still alive. Then the breaker-bytes get removed automatically by aggregator close. But, they are still in use until the object goes away.

I am not sure how this all plays together with 
```
        this.mapReduceContextByBucketOrdinal = new LongObjectPagedHashMap<>(1, context.bigArrays());
```
and 
```
      bigArraysForMapReduce
```

Does calling `close` on `mapReduceContextByBucketOrdinal` call close on all its objects? Or just on the hashmap collection?

And each internal agg is keeping a reference to a `TransactionStore` that the aggregator has a reference to as well (at least until serialization, then I suppose the reference is dropped). This seems pretty fragile. Basically, the aggregator and the internal aggregations are references the same object instances and nothing is guaranteeing one does not mutate the other.

Reply:
> Does calling close on mapReduceContextByBucketOrdinal call close on all its objects? Or just on the hashmap collection?

No, it is not that smart. I tried the other variant first: deep copy every context. In this variant I had to loop over all objects in the `LongObjectPagedHashMap` and close them.

The closing part works, at least manual testing that thing even on a single-node cluster produces a SearchPhaseExecutionException if the item gets closed in the aggregator but is till used in the internal agg. 

> This seems pretty fragile. Basically, the aggregator and the internal aggregations are references the same object instances and nothing is guaranteeing one does not mutate the other.

100% agree. What will help: I am working on a `SealedTransactionStore` (or should I call it ImmutableTransactionStore?) which will implement the idea of dropping the hashtable and only keep the strings and counts. It will not allow mutating transactions in the internal agg as a nice side effect.

I am not happy how all the CB magic works. The problem is the assumptions we make today aren't necessarily given tomorrow when search/aggs changes something. Hopefully a multi-node test will cover use.